crypto/timestamp/timestamptest/tsa.go

```diff
@@ -0,0 +1,269 @@
+}
+
+// Timestamp stamps the time with the given request.
+func (tsa *TSA) Timestamp(_ context.Context, req *timestamp.Request) (*timestamp.Response, error) {
+	// validate request
+	if req.Version != 1 {
+		return responseRejection, nil
+	}
+	hash, ok := oid.ConvertToHash(req.MessageImprint.HashAlgorithm.Algorithm)
+	if !ok {
+		return responseRejection, nil
+	}
+	if hashedMessage := req.MessageImprint.HashedMessage; len(hashedMessage) != hash.Size() {
+		return responseRejection, nil
+	}
+
+	// generate token info
+	policy := asn1.ObjectIdentifier{1, 3, 6, 1, 4, 1, 4146, 2} // time-stamp-policies
+	switch hash {
+	case crypto.SHA1:
+		policy = append(policy, 2)
+	case crypto.SHA256, crypto.SHA384, crypto.SHA512:
+		policy = append(policy, 3)
+	default:
+		return responseRejection, nil
+	}
+	infoBytes, err := tsa.generateTokenInfo(req, policy)
+	if err != nil {
+		return nil, err
+	}
+
+	// generate signed data
+	signed, err := tsa.generateSignedData(infoBytes, req.CertReq)
+	if err != nil {
+		return nil, err
+	}
+	content, err := convertToRawASN1(signed, "explicit,tag:0")
+	if err != nil {
+		return nil, err
+	}
+
+	// generate content info
+	contentInfo := cms.ContentInfo{
+		ContentType: oid.SignedData,
+		Content:     content,
+	}
+	token, err := convertToRawASN1(contentInfo, "")
+	if err != nil {
+		return nil, err
+	}
+
+	// generate response
+	return &timestamp.Response{
+		Status: pki.StatusInfo{
+			Status: pki.StatusGranted,
+		},
+		TimeStampToken: token,
+	}, nil
+}
+
+// generateTokenInfo generate timestamp token info.
+func (tsa *TSA) generateTokenInfo(req *timestamp.Request, policy asn1.ObjectIdentifier) ([]byte, error) {
+	serialNumber, err := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+	if err != nil {
+		return nil, err
+	}
+	nowFunc := tsa.NowFunc
+	if nowFunc == nil {
+		nowFunc = time.Now
+	}
+	info := timestamp.TSTInfo{
+		Version:        1,
+		Policy:         policy,
+		MessageImprint: req.MessageImprint,
+		SerialNumber:   serialNumber,
+		GenTime:        nowFunc().UTC().Truncate(time.Second),
+		Accuracy: timestamp.Accuracy{
+			Seconds: 1,
+		},
+	}
+	return asn1.Marshal(info)
+}
+
+// generateSignedData generate signed data according to
+func (tsa *TSA) generateSignedData(infoBytes []byte, requestCert bool) (cms.SignedData, error) {
+	var issuer asn1.RawValue
+	_, err := asn1.Unmarshal(tsa.cert.RawIssuer, &issuer)
+	if err != nil {
+		return cms.SignedData{}, err
+	}
+	contentType, err := convertToRawASN1([]interface{}{oid.TSTInfo}, "set")
+	if err != nil {
+		return cms.SignedData{}, err
+	}
+	infoDigest, err := hashutil.ComputeHash(crypto.SHA256, infoBytes)
+	if err != nil {
+		return cms.SignedData{}, err
+	}
+	messageDigest, err := convertToRawASN1([]interface{}{infoDigest}, "set")
+	if err != nil {
+		return cms.SignedData{}, err
+	}
+	signingTime, err := convertToRawASN1([]interface{}{time.Now().UTC()}, "set")
+	if err != nil {
+		return cms.SignedData{}, err
+	}
+	signed := cms.SignedData{
+		Version: 3,
```

Comment:
Is the version defined somewhere?

Reply:
See [RFC 5652 5.1](https://datatracker.ietf.org/doc/html/rfc5652#section-5.1). More info can be found in the `internal/crypto/cms` package.